### DISCUSSION TITLE
Porting MiST cores to Tang nano 20k

### DISCUSSION THREAD

**Discussion by somhi** [nand2mario/nestang#17] [General]
I have Apple 2e half ported from MiST to tang nano 20k with [DeMiSTify framework](https://github.com/robinsonb5/DeMiSTify)

[https://github.com/somhi/gowin_stuff/tree/main/Apple2efpga/tangnano20k](https://github.com/somhi/gowin_stuff/tree/main/Apple2efpga/tangnano20k)

What works: video, SD card (I'm able to see files from OSD).
VGA video is Ok. HDMI video needs to be corrected, but is enough for seeing if SDRAM is working.
When SDRAM is working, a cursor has to be seen at the bottom of the screen, or an Apple // text on top of it.  
If a disk is loaded (e.g. nib files), a cold reset is needed thereafter.

SDRAM controller is what is needed to be adapted in this core.  
Don't have enough knowledge to adapt it. 

Haven't seen any SDRAM controller working apart from yours.  
Is there any way to adapt "standard" MiST/MiSTer controllers to work with tang nano 20k?
If for every ported core is needed to get your SDRAM controller and adapt to each core, I don't think it will be an easy and fast task.

I won't touch anything else from this core for now until I see more people succeeding with getting SDRAM working.

**Comment by nand2mario** [comment]
Great to see an apple 2 core. The SDRAM controller is actually quite straightforward to use. Let me prepare some documentation. Later we can also look at compatibility with mister api.

**Comment by nand2mario** [comment]
I've posted an example of using the SDRAM controller: https://github.com/nand2mario/sdram-tang-nano-20k/tree/main. Let me know if it helps.

**Comment by somhi** [comment]
Thanks, that is a great example on using your SDRAM controller on new cores. I've tested it and works great.

The thing is that each core already done will have its own SDRAM controller.  Changing them for your controller needs a great knowledge, which I don't have it right now.   If we could instead use the controllers already done for MiST & MiSTer, changing the relevant parameters, it will be easier to port cores.

**Comment by nand2mario** [comment]
> Thanks, that is a great example on using your SDRAM controller on new cores. I've tested it and works great.
> 
> 
> 
> The thing is that each core already done will have its own SDRAM controller.  Changing them for your controller needs a great knowledge, which I don't have it right now.   If we could instead use the controllers already done for MiST & MiSTer, changing the relevant parameters, it will be easier to port cores. 
> 
> 
Right. That would be the right solution in the future.
